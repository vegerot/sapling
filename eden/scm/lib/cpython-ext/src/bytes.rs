/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use cpython::*;

#[derive(Clone, Debug, Eq, PartialEq, PartialOrd, Default, Hash, Ord)]
pub struct Bytes(pub Box<[u8]>);

impl ToPyObject for Bytes {
    type ObjectType = PyBytes;

    #[inline]
    fn to_py_object(&self, py: Python) -> PyBytes {
        PyBytes::new(py, &self.0)
    }
}

impl<'source> FromPyObject<'source> for Bytes {
    fn extract(py: Python, obj: &'source PyObject) -> PyResult<Self> {
        let data = obj.cast_as::<PyBytes>(py)?.data(py);
        Ok(Bytes(data.to_vec().into_boxed_slice()))
    }
}

impl From<Box<[u8]>> for Bytes {
    fn from(v: Box<[u8]>) -> Bytes {
        Bytes(v)
    }
}

impl From<Vec<u8>> for Bytes {
    fn from(v: Vec<u8>) -> Bytes {
        Bytes(v.into_boxed_slice())
    }
}

impl From<String> for Bytes {
    fn from(s: String) -> Bytes {
        Bytes(s.into_bytes().into_boxed_slice())
    }
}

impl AsRef<[u8]> for Bytes {
    fn as_ref(&self) -> &[u8] {
        &self.0
    }
}
