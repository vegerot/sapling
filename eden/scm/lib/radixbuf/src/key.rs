/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This software may be used and distributed according to the terms of the
 * GNU General Public License version 2.
 */

//! Types and helper functions to convert a `KeyId` to a Key (`[u8]`). The radix index is
//! responsible for the other direction - `[u8]` to `KeyId`. This mapping is to remove the
//! need of storing full keys in the radix index.
//!
//! Typically, there is an append-only file which is the source of truth of some information.
//! The `KeyId` could be offsets in that file.
//!
//! The radix index APIs only care about how to read a key, given a `KeyId`. It does not
//! care about how to write a key. This is more flexible. For example, the source of truth
//! file may have keys stored already before the index gets built. The `append` methods provided
//! below are only for convenience.
//!
//! There could be cases where the `KeyId`s are not offsets, or the keys are not stored as-is.
//! For example, `KeyId` could be mercurial revision numbers instead of offsets for non-inlined
//! revlog .i files. The radix index has a limitation (to make the format slightly more compact)
//! that one key cannot be the prefix of another different key. This could be guaranteed by using
//! the fixed length key, or appending `'\0'` to strings that cannot have `'\0'`. But in case
//! of variant-length binary key, some escaping is needed to avoid the prefix conflict. The
//! escaping logic might be implemented as key read/write functions.

use std::io::Write;

use vlqencoding::VLQDecodeAt;
use vlqencoding::VLQEncode;

use crate::errors::ErrorKind;
use crate::traits::Resize;
use crate::Result;

/// Integer that maps to a key (`[u8]`).
#[derive(Debug, Eq, PartialEq, PartialOrd, Ord, Clone, Copy, Default)]
pub struct KeyId(u32);

macro_rules! impl_convert {
    ($T: ty) => {
        impl From<$T> for KeyId {
            #[allow(unused_comparisons)]
            #[inline]
            fn from(v: $T) -> Self {
                if v > 0xffff_ffff || v < 0 {
                    panic!("KeyId out of range")
                }
                KeyId(v as u32)
            }
        }

        impl Into<$T> for KeyId {
            #[inline]
            fn into(self) -> $T {
                self.0 as $T
            }
        }
    };
}

impl_convert!(u32);
impl_convert!(u64);

// Make sure `usize` contains at least 4 bytes (`KeyId` size).
const _SIZE_TEST: usize = 0xffff_ffff;
impl_convert!(usize);

/// Keys with fixed 20 bytes length.
pub struct FixedKey;

/// Keys with variant-length. Serialized as a VLQ-encoded length, followed by the actual key.
pub struct VariantKey;

impl FixedKey {
    #[inline]
    pub fn read<'a, K: AsRef<[u8]>>(key_buf: &'a K, key_id: KeyId) -> Result<&'a [u8]> {
        let key_buf = key_buf.as_ref();
        let start_pos: usize = key_id.into();
        // LANG: Consider making 20 a type parameter once supported.
        let end_pos = start_pos + 20;
        if key_buf.len() < end_pos {
            return Err(ErrorKind::InvalidKeyId(key_id));
        }
        Ok(&key_buf[start_pos..end_pos])
    }

    #[inline]
    pub fn append<B: AsMut<[u8]> + Resize<u8>, K: AsRef<[u8]>>(key_buf: &mut B, key: &K) -> KeyId {
        let key = key.as_ref();
        assert_eq!(key.len(), 20);

        let pos = key_buf.as_mut().len();
        key_buf.resize(pos + 20, 0);
        let key_buf = key_buf.as_mut();
        key_buf[pos..pos + 20].copy_from_slice(key);
        pos.into()
    }
}

impl VariantKey {
    #[inline]
    pub fn read<'a, K: AsRef<[u8]>>(key_buf: &'a K, key_id: KeyId) -> Result<&'a [u8]> {
        let key_buf = key_buf.as_ref();
        let (key_len, vlq_size): (usize, _) = key_buf
            .read_vlq_at(key_id.0 as usize)
            .map_err(|_| ErrorKind::InvalidKeyId(key_id))?;
        let start_pos = key_id.0 as usize + vlq_size;
        let end_pos = start_pos + key_len;
        if key_buf.len() < end_pos {
            return Err(ErrorKind::InvalidKeyId(key_id));
        }
        Ok(&key_buf[start_pos..end_pos])
    }

    #[inline]
    pub fn append<B: AsMut<[u8]> + Resize<u8>, K: AsRef<[u8]>>(key_buf: &mut B, key: &K) -> KeyId {
        let key = key.as_ref();
        // PERF: the Vec allocation may be avoided with a more complex implementation.
        // Most of the time, key length could be encoded in 2-byte VLQ. Pre-allocate 2 bytes.
        let mut buf = Vec::<u8>::with_capacity(key.len() + 2);
        buf.write_vlq(key.len()).expect("write len");
        buf.write_all(key).expect("write key");

        let pos = key_buf.as_mut().len();
        key_buf.resize(pos + buf.len(), 0);
        let key_buf = key_buf.as_mut();
        key_buf[pos..pos + buf.len()].copy_from_slice(&buf[..]);
        (pos as u64).into()
    }
}

#[cfg(test)]
mod tests {
    use rand::distributions::Alphanumeric;
    use rand::thread_rng;
    use rand::Rng;
    use rand::RngCore;

    use super::*;

    #[test]
    fn test_variant_key_round_trip() {
        let mut buf = Vec::<u8>::new();
        let keys: Vec<Vec<u8>> = (0..1000usize)
            .map(|i| {
                let rng = thread_rng();
                rng.sample_iter(&Alphanumeric).take(i % 40).collect()
            })
            .collect();
        let key_ids: Vec<KeyId> = keys
            .iter()
            .map(|key| VariantKey::append(&mut buf, key))
            .collect();
        let keys_retrieved: Vec<Vec<u8>> = key_ids
            .iter()
            .map(|&id| VariantKey::read(&buf, id).unwrap().to_vec())
            .collect();
        assert_eq!(keys, keys_retrieved);
    }

    #[test]
    fn test_fixed_key_round_trip() {
        let mut rng = thread_rng();
        let mut buf = Vec::<u8>::new();
        let keys: Vec<[u8; 20]> = (0..1000usize)
            .map(|_| {
                let mut bytes = [0u8; 20];
                rng.fill_bytes(&mut bytes);
                bytes
            })
            .collect();
        let key_ids: Vec<KeyId> = keys
            .iter()
            .map(|key| FixedKey::append(&mut buf, key))
            .collect();
        assert_eq!(buf.len(), 1000 * 20);
        let keys_retrieved: Vec<[u8; 20]> = key_ids
            .iter()
            .map(|&id| {
                let mut bytes = [0u8; 20];
                bytes.copy_from_slice(FixedKey::read(&buf, id).unwrap());
                bytes
            })
            .collect();
        assert_eq!(keys, keys_retrieved);
    }
}
