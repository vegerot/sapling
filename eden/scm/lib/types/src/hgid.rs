/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This software may be used and distributed according to the terms of the
 * GNU General Public License version 2.
 */

#[cfg(any(test, feature = "for-tests"))]
use std::collections::HashSet;
use std::io;
use std::io::Read;
use std::io::Write;

#[cfg(any(test, feature = "for-tests"))]
use rand::RngCore;
use sha1::Digest;
use sha1::Sha1;

use crate::hash::AbstractHashType;
use crate::hash::HashTypeInfo;
use crate::parents::Parents;

/// A 20-byte identifier, often a hash. Nodes are used to uniquely identify
/// commits, file versions, and many other things.
///
///
/// # Serde Serialization
///
/// The `serde_with` module allows customization on `HgId` serialization:
/// - `#[serde(with = "types::serde_with::hgid::bytes")]` (current default)
/// - `#[serde(with = "types::serde_with::hgid::hex")]`
/// - `#[serde(with = "types::serde_with::hgid::tuple")]`
///
/// Using them can change the size or the type of serialization result:
///
/// | lib \ serde_with | hgid::tuple         | hgid::bytes  | hgid::hex |
/// |------------------|---------------------|--------------|-----------|
/// | mincode          | 20 bytes            | 21 bytes     | 41 bytes  |
/// | cbor             | 21 to 41 bytes  [1] | 21 bytes     | 42 bytes  |
/// | json             | 41 to 81+ bytes [1] | invalid  [2] | 42 bytes  |
/// | python           | Tuple[int]          | bytes        | str       |
///
/// In general,
/// - `hgid::tuple` only works best for `mincode`.
/// - `hgid::bytes` works best for cbor, python.
/// - `hgid::hex` is useful for `json`, or other text-only formats.
///
/// Compatibility note:
/// - `hgid::tuple` cannot decode `hgid::bytes` or `hgid::hex` data.
/// - `hgid::hex` can decode `hgid::bytes` data, or vice-versa. They share a
///   same `deserialize` implementation.
/// - `hgid::hex` or `hgid::bytes` might be able to decode `hgid::tuple` data,
///   depending on how tuples are serialized. For example, mincode
///   does not add framing for tuples, so `hgid::bytes` cannot decode
///   `hgid::tuple` data; cbor adds framing for tuples, so `hgid::bytes`
///   can decode `hgid::tuple` data.
///
/// [1]: Depends on actual data of `HgId`.
/// [2]: JSON only supports utf-8 data.
pub type HgId = AbstractHashType<HgIdTypeInfo, 20>;

pub struct HgIdTypeInfo;

impl HashTypeInfo for HgIdTypeInfo {
    const HASH_TYPE_NAME: &'static str = "HgId";
}

/// The nullid (0x00) is used throughout Mercurial to represent "None".
/// (For example, a commit will have a nullid p2, if it has no second parent).
pub const NULL_ID: HgId = HgId::from_byte_array([0; HgId::len()]);

/// The hard-coded 'working copy parent' Mercurial id.
pub const WDIR_ID: HgId = HgId::from_byte_array([0xff; HgId::len()]);

pub const WDIR_REV: i64 = i64::MAX;

/// The hard-coded 'pending file add' manifest node id.
pub const MF_ADDED_NODE_ID: HgId = HgId::from_byte_array([
    b'0', b'0', b'0', b'0', b'0', b'0', b'0', b'0', b'0', b'0', b'0', b'0', b'0', b'0', b'0', b'a',
    b'd', b'd', b'e', b'd',
]);

/// The hard-coded 'pending file modification' manifest node id.
pub const MF_MODIFIED_NODE_ID: HgId = HgId::from_byte_array([
    b'0', b'0', b'0', b'0', b'0', b'0', b'0', b'0', b'0', b'0', b'0', b'0', b'm', b'o', b'd', b'i',
    b'f', b'i', b'e', b'd',
]);

/// The hard-coded 'untracked file' manifest node id.
pub const MF_UNTRACKED_NODE_ID: HgId = HgId::from_byte_array([
    b'0', b'0', b'0', b'0', b'0', b'0', b'0', b'0', b'0', b'0', b'0', b'u', b'n', b't', b'r', b'a',
    b'c', b'k', b'e', b'd',
]);

impl HgId {
    pub const fn null_id() -> &'static Self {
        &NULL_ID
    }

    pub fn is_null(&self) -> bool {
        self == &NULL_ID
    }

    pub const fn wdir_id() -> &'static Self {
        &WDIR_ID
    }

    pub fn is_wdir(&self) -> bool {
        self == &WDIR_ID
    }

    pub fn from_content(data: &[u8], parents: Parents) -> Self {
        // Parents must be hashed in sorted order.
        let (p1, p2) = match parents.into_nodes() {
            (p1, p2) if p1 > p2 => (p2, p1),
            (p1, p2) => (p1, p2),
        };

        let mut hasher = Sha1::new();
        hasher.update(p1.as_ref());
        hasher.update(p2.as_ref());
        hasher.update(data);
        let hash: [u8; 20] = hasher.finalize().into();

        HgId::from_byte_array(hash)
    }

    #[cfg(any(test, feature = "for-tests"))]
    pub fn random(rng: &mut dyn RngCore) -> Self {
        let mut bytes = [0; HgId::len()];
        rng.fill_bytes(&mut bytes);
        loop {
            let hgid = HgId::from(&bytes);
            if !hgid.is_null() {
                return hgid;
            }
        }
    }

    #[cfg(any(test, feature = "for-tests"))]
    pub fn random_distinct(rng: &mut dyn RngCore, count: usize) -> Vec<Self> {
        let mut nodes = Vec::new();
        let mut nodeset = HashSet::new();
        while nodes.len() < count {
            let hgid = HgId::random(rng);
            if !nodeset.contains(&hgid) {
                nodeset.insert(hgid.clone());
                nodes.push(hgid);
            }
        }
        nodes
    }
}

impl<'a> From<&'a [u8; HgId::len()]> for HgId {
    fn from(bytes: &[u8; HgId::len()]) -> HgId {
        HgId::from_byte_array(bytes.clone())
    }
}

pub trait WriteHgIdExt {
    /// Write a ``HgId`` directly to a stream.
    ///
    /// # Examples
    ///
    /// ```
    /// use types::hgid::HgId;
    /// use types::hgid::WriteHgIdExt;
    /// let mut v = vec![];
    ///
    /// let n = HgId::null_id();
    /// v.write_hgid(&n)
    ///     .expect("writing a hgid to a vec should work");
    ///
    /// assert_eq!(v, vec![0; HgId::len()]);
    /// ```
    fn write_hgid(&mut self, value: &HgId) -> io::Result<()>;
}

impl<W: Write + ?Sized> WriteHgIdExt for W {
    fn write_hgid(&mut self, value: &HgId) -> io::Result<()> {
        self.write_all(value.as_ref())
    }
}

pub trait ReadHgIdExt {
    /// Read a ``HgId`` directly from a stream.
    ///
    /// # Examples
    ///
    /// ```
    /// use std::io::Cursor;
    ///
    /// use types::hgid::HgId;
    /// use types::hgid::ReadHgIdExt;
    /// let mut v = vec![0; HgId::len()];
    /// let mut c = Cursor::new(v);
    ///
    /// let n = c
    ///     .read_hgid()
    ///     .expect("reading a hgid from a vec should work");
    ///
    /// assert_eq!(&n, HgId::null_id());
    /// ```
    fn read_hgid(&mut self) -> io::Result<HgId>;
}

impl<R: Read + ?Sized> ReadHgIdExt for R {
    fn read_hgid(&mut self) -> io::Result<HgId> {
        let mut bytes = [0; HgId::len()];
        self.read_exact(&mut bytes)?;
        Ok(HgId::from_byte_array(bytes))
    }
}

#[cfg(any(test, feature = "for-tests"))]
pub mod mocks {
    use super::HgId;

    pub const ONES: HgId = HgId::from_byte_array([0x11; HgId::len()]);
    pub const TWOS: HgId = HgId::from_byte_array([0x22; HgId::len()]);
    pub const THREES: HgId = HgId::from_byte_array([0x33; HgId::len()]);
    pub const FOURS: HgId = HgId::from_byte_array([0x44; HgId::len()]);
    pub const FIVES: HgId = HgId::from_byte_array([0x55; HgId::len()]);
    pub const SIXES: HgId = HgId::from_byte_array([0x66; HgId::len()]);
    pub const SEVENS: HgId = HgId::from_byte_array([0x77; HgId::len()]);
    pub const EIGHTS: HgId = HgId::from_byte_array([0x88; HgId::len()]);
    pub const NINES: HgId = HgId::from_byte_array([0x99; HgId::len()]);
    pub const AS: HgId = HgId::from_byte_array([0xAA; HgId::len()]);
    pub const BS: HgId = HgId::from_byte_array([0xAB; HgId::len()]);
    pub const CS: HgId = HgId::from_byte_array([0xCC; HgId::len()]);
    pub const DS: HgId = HgId::from_byte_array([0xDD; HgId::len()]);
    pub const ES: HgId = HgId::from_byte_array([0xEE; HgId::len()]);
    pub const FS: HgId = HgId::from_byte_array([0xFF; HgId::len()]);
}

#[cfg(test)]
mod tests {
    use quickcheck::quickcheck;
    use serde::Deserialize;
    use serde::Serialize;

    use super::*;

    #[test]
    fn test_incorrect_length() {
        HgId::from_slice(&[0u8; 25]).expect_err("bad slice length");
    }

    #[test]
    fn test_serde_with_using_cbor() {
        // Note: this test is for CBOR. Other serializers like mincode
        // or Thrift would have different backwards compatibility!
        use serde_cbor::de::from_slice as decode;
        use serde_cbor::ser::to_vec as encode;

        #[derive(Serialize, Deserialize, Debug, Eq, PartialEq)]
        struct Orig(#[serde(with = "crate::serde_with::hgid::tuple")] HgId);

        #[derive(Serialize, Deserialize, Debug, Eq, PartialEq)]
        struct Bytes(#[serde(with = "crate::serde_with::hgid::bytes")] HgId);

        #[derive(Serialize, Deserialize, Debug, Eq, PartialEq)]
        struct Hex(#[serde(with = "crate::serde_with::hgid::hex")] HgId);

        let id: HgId = mocks::CS;
        let orig = Orig(id);
        let bytes = Bytes(id);
        let hex = Hex(id);

        let cbor_orig = encode(&orig).unwrap();
        let cbor_bytes = encode(&bytes).unwrap();
        let cbor_hex = encode(&hex).unwrap();

        assert_eq!(cbor_orig.len(), 41);
        assert_eq!(cbor_bytes.len(), 21);
        assert_eq!(cbor_hex.len(), 42);

        // Orig cannot decode bytes or hex.
        assert_eq!(decode::<Orig>(&cbor_orig).unwrap().0, id);
        decode::<Orig>(&cbor_bytes).unwrap_err();
        decode::<Orig>(&cbor_hex).unwrap_err();

        // Bytes can decode all 3 formats.
        assert_eq!(decode::<Bytes>(&cbor_orig).unwrap().0, id);
        assert_eq!(decode::<Bytes>(&cbor_bytes).unwrap().0, id);
        assert_eq!(decode::<Bytes>(&cbor_hex).unwrap().0, id);

        // Hex can decode all 3 formats.
        assert_eq!(decode::<Hex>(&cbor_orig).unwrap().0, id);
        assert_eq!(decode::<Hex>(&cbor_bytes).unwrap().0, id);
        assert_eq!(decode::<Hex>(&cbor_hex).unwrap().0, id);
    }

    quickcheck! {
        fn test_from_slice(hgid: HgId) -> bool {
            hgid == HgId::from_slice(hgid.as_ref()).expect("from_slice")
        }
    }
}
