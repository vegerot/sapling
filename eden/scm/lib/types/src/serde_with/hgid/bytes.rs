/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use serde::de::Error;
use serde::Deserializer;
use serde::Serializer;

use crate::HgId;

/// Serde `serialize_with` function to serialize `HgId` as bytes.
pub fn serialize<S>(id: &HgId, serializer: S) -> Result<S::Ok, S::Error>
where
    S: Serializer,
{
    serializer.serialize_bytes(id.as_ref())
}

/// Serde `deserialize_with` function to deserialize `HgId` as list of u8s,
/// or bytes, or a hex string.
pub fn deserialize<'de, D>(deserializer: D) -> Result<HgId, D::Error>
where
    D: Deserializer<'de>,
{
    // ByteBuf supports both list and bytes.
    let bytes: serde_bytes::ByteBuf = serde_bytes::deserialize(deserializer)?;
    let bytes = bytes.as_ref();
    // Compatible with hex.
    if bytes.len() == HgId::hex_len() {
        HgId::from_hex(bytes).map_err(|e| {
            let msg = format!("invalid HgId: {} ({:?})", e, bytes);
            D::Error::custom(msg)
        })
    } else {
        HgId::from_slice(bytes).map_err(|e| {
            let msg = format!("invalid HgId: {} ({:?})", e, bytes);
            D::Error::custom(msg)
        })
    }
}
