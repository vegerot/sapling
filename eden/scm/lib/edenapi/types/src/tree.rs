/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This software may be used and distributed according to the terms of the
 * GNU General Public License version 2.
 */

use bytes::Bytes;
#[cfg(any(test, feature = "for-tests"))]
use quickcheck::Arbitrary;
#[cfg(any(test, feature = "for-tests"))]
use quickcheck_arbitrary_derive::Arbitrary;
use serde_derive::Deserialize;
use serde_derive::Serialize;
use thiserror::Error;
use type_macros::auto_wire;
use types::hgid::HgId;
use types::key::Key;
use types::parents::Parents;

use crate::DirectoryMetadata;
use crate::EdenApiServerError;
use crate::FileMetadata;
use crate::InvalidHgId;
use crate::UploadToken;

#[derive(Debug, Error)]
pub enum TreeError {
    #[error(transparent)]
    Corrupt(InvalidHgId),
    /// If this entry represents a root node (i.e., has an empty path), it
    /// may be a hybrid tree manifest which has the content of a root tree
    /// manifest node, but the hash of the corresponding flat manifest. This
    /// situation should only occur for manifests created in "hybrid mode"
    /// (i.e., during a transition from flat manifests to tree manifests).
    #[error("Detected possible hybrid manifest: {0}")]
    MaybeHybridManifest(#[source] InvalidHgId),

    #[error("TreeEntry missing field '{0}'")]
    MissingField(&'static str),
}

impl TreeError {
    /// Get the data anyway, despite the error.
    pub fn data(&self) -> Option<Bytes> {
        use TreeError::*;
        match self {
            Corrupt(InvalidHgId { data, .. }) => Some(data),
            MaybeHybridManifest(InvalidHgId { data, .. }) => Some(data),
            _ => None,
        }
        .cloned()
    }
}

/// Structure representing source control tree entry on the wire.
/// Includes the information required to add the data to a mutable store,
/// along with the parents for hash validation.
#[derive(Clone, Debug, Default, Eq, PartialEq, Deserialize, Serialize)]
pub struct TreeEntry {
    pub key: Key,
    pub data: Option<Bytes>,
    pub parents: Option<Parents>,
    #[serde(skip)]
    pub children: Option<Vec<Result<TreeChildEntry, EdenApiServerError>>>,
}

impl TreeEntry {
    pub fn new(key: Key, data: Bytes, parents: Parents) -> Self {
        Self {
            key,
            data: Some(data),
            parents: Some(parents),
            children: None,
        }
    }

    pub fn with_children<'a>(
        &'a mut self,
        children: Option<Vec<Result<TreeChildEntry, EdenApiServerError>>>,
    ) -> &'a mut Self {
        self.children = children;
        self
    }

    pub fn key(&self) -> &Key {
        &self.key
    }

    /// Get this entry's data. Checks data integrity but allows hash mismatches
    /// if this is a suspected hybrid manifest.
    pub fn data(&self) -> Result<Bytes, TreeError> {
        use TreeError::*;
        self.data_checked().or_else(|e| match e {
            MaybeHybridManifest(_) => Ok(e
                .data()
                .expect("TreeError::MaybeHybridManifest should always carry underlying data")),
            _ => Err(e),
        })
    }

    /// Get this entry's data after verifying the hgid hash.
    pub fn data_checked(&self) -> Result<Bytes, TreeError> {
        if let Some(data) = self.data.as_ref() {
            if let Some(parents) = self.parents {
                let computed = HgId::from_content(data, parents);
                if computed != self.key.hgid {
                    let err = InvalidHgId {
                        expected: self.key.hgid,
                        computed,
                        data: data.clone(),
                        parents,
                    };

                    return Err(if self.key.path.is_empty() {
                        TreeError::MaybeHybridManifest(err)
                    } else {
                        TreeError::Corrupt(err)
                    });
                }

                Ok(data.clone())
            } else {
                Err(TreeError::MissingField("parents"))
            }
        } else {
            Err(TreeError::MissingField("data"))
        }
    }

    /// Get this entry's data without verifying the hgid hash.
    pub fn data_unchecked(&self) -> Option<Bytes> {
        self.data.clone()
    }
}

#[derive(Clone, Debug, Eq, PartialEq)]
#[cfg_attr(any(test, feature = "for-tests"), derive(Arbitrary))]
pub enum TreeChildEntry {
    File(TreeChildFileEntry),
    Directory(TreeChildDirectoryEntry),
}

#[derive(Clone, Debug, Default, Eq, PartialEq)]
#[cfg_attr(any(test, feature = "for-tests"), derive(Arbitrary))]
pub struct TreeChildFileEntry {
    pub key: Key,
    pub file_metadata: Option<FileMetadata>,
}

#[derive(Clone, Debug, Default, Eq, PartialEq)]
#[cfg_attr(any(test, feature = "for-tests"), derive(Arbitrary))]
pub struct TreeChildDirectoryEntry {
    pub key: Key,
    pub directory_metadata: Option<DirectoryMetadata>,
}

impl TreeChildEntry {
    pub fn new_file_entry(key: Key, metadata: FileMetadata) -> Self {
        TreeChildEntry::File(TreeChildFileEntry {
            key,
            file_metadata: Some(metadata),
        })
    }

    pub fn new_directory_entry(key: Key) -> Self {
        TreeChildEntry::Directory(TreeChildDirectoryEntry {
            key,
            ..Default::default()
        })
    }
}

#[cfg(any(test, feature = "for-tests"))]
impl Arbitrary for TreeEntry {
    fn arbitrary(g: &mut quickcheck::Gen) -> Self {
        let bytes: Option<Vec<u8>> = Arbitrary::arbitrary(g);
        Self {
            key: Arbitrary::arbitrary(g),
            data: bytes.map(Bytes::from),
            parents: Arbitrary::arbitrary(g),
            // Recursive TreeEntry in children causes stack overflow in QuickCheck
            children: None,
        }
    }
}

#[derive(Clone, Debug, Default, Deserialize, Serialize, Eq, PartialEq)]
#[cfg_attr(any(test, feature = "for-tests"), derive(Arbitrary))]
pub struct TreeRequest {
    pub keys: Vec<Key>,
    pub attributes: TreeAttributes,
}

#[derive(Copy, Clone, Debug, Deserialize, Serialize, Eq, PartialEq)]
#[cfg_attr(any(test, feature = "for-tests"), derive(Arbitrary))]
pub struct TreeAttributes {
    #[serde(default = "get_true")]
    pub manifest_blob: bool,
    #[serde(default = "get_true")]
    pub parents: bool,
    #[serde(default = "get_true")]
    pub child_metadata: bool,
}

fn get_true() -> bool {
    true
}

impl TreeAttributes {
    pub fn all() -> Self {
        TreeAttributes {
            manifest_blob: true,
            parents: true,
            child_metadata: true,
        }
    }
}

impl Default for TreeAttributes {
    fn default() -> Self {
        TreeAttributes {
            manifest_blob: true,
            parents: true,
            child_metadata: false,
        }
    }
}

#[auto_wire]
#[derive(Clone, Debug, Default, Eq, PartialEq, Serialize, Deserialize)]
#[cfg_attr(any(test, feature = "for-tests"), derive(Arbitrary))]
pub struct UploadTreeEntry {
    #[id(0)]
    pub node_id: HgId,
    #[id(1)]
    pub data: Vec<u8>,
    #[id(2)]
    pub parents: Parents,
}

#[auto_wire]
#[derive(Clone, Default, Debug, Deserialize, Serialize, Eq, PartialEq)]
#[cfg_attr(any(test, feature = "for-tests"), derive(Arbitrary))]
pub struct UploadTreeRequest {
    #[id(0)]
    pub entry: UploadTreeEntry,
}

#[auto_wire]
#[derive(Clone, Default, Debug, Deserialize, Serialize, Eq, PartialEq)]
#[cfg_attr(any(test, feature = "for-tests"), derive(Arbitrary))]
pub struct UploadTreeResponse {
    #[id(1)]
    pub token: UploadToken,
}
