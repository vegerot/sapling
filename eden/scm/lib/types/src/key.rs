/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use std::fmt;

use serde_derive::Deserialize;
use serde_derive::Serialize;

use crate::hgid::HgId;
use crate::path::RepoPathBuf;

#[derive(
    Clone,
    Debug,
    Default,
    Eq,
    Hash,
    Ord,
    PartialEq,
    PartialOrd,
    Serialize,
    Deserialize
)]
#[cfg_attr(
    any(test, feature = "for-tests"),
    derive(quickcheck_arbitrary_derive::Arbitrary)
)]
pub struct Key {
    // Name is usually a file or directory path
    pub path: RepoPathBuf,
    // HgId is always a 20 byte hash. This will be changed to a fix length array later.
    #[serde(alias = "node")]
    #[serde(rename = "node")]
    pub hgid: HgId,
}

impl Key {
    pub fn new(path: RepoPathBuf, hgid: HgId) -> Self {
        Key { path, hgid }
    }
}

impl fmt::Display for Key {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{} {}", &self.hgid, self.path)
    }
}

#[cfg(any(test, feature = "for-tests"))]
pub mod mocks {
    use lazy_static::lazy_static;

    use super::Key;
    use crate::hgid::mocks::ONES;
    use crate::hgid::mocks::THREES;
    use crate::hgid::mocks::TWOS;
    use crate::testutil::*;

    lazy_static! {
        pub static ref FOO_KEY: Key = Key::new(repo_path_buf("foo"), ONES);
        pub static ref BAR_KEY: Key = Key::new(repo_path_buf("bar"), TWOS);
        pub static ref BAZ_KEY: Key = Key::new(repo_path_buf("baz"), THREES);
    }
}

#[cfg(test)]
mod tests {
    use mocks::*;

    use super::*;

    #[test]
    fn display_key() {
        let foo = "1111111111111111111111111111111111111111 foo";
        let bar = "2222222222222222222222222222222222222222 bar";
        let baz = "3333333333333333333333333333333333333333 baz";
        assert_eq!(format!("{}", &*FOO_KEY), foo);
        assert_eq!(format!("{}", &*BAR_KEY), bar);
        assert_eq!(format!("{}", &*BAZ_KEY), baz);
    }

    #[test]
    fn test_serde_with_using_cbor() {
        // Note: this test is for CBOR. Other serializers like mincode
        // or Thrift would have different backwards compatibility!
        use serde_cbor::de::from_slice as decode;
        use serde_cbor::ser::to_vec as encode;

        #[derive(Serialize, Deserialize, Debug, Eq, PartialEq)]
        struct Orig(
            #[serde(with = "crate::serde_with::key::tuple")] Key,
            #[serde(with = "crate::serde_with::key::tuple")] Option<Key>,
        );

        #[derive(Serialize, Deserialize, Debug, Eq, PartialEq)]
        struct Bytes(
            #[serde(with = "crate::serde_with::key::bytes")] Key,
            #[serde(with = "crate::serde_with::key::bytes")] Option<Key>,
        );

        #[derive(Serialize, Deserialize, Debug, Eq, PartialEq)]
        struct Hex(
            #[serde(with = "crate::serde_with::key::hex")] Key,
            #[serde(with = "crate::serde_with::key::hex")] Option<Key>,
        );

        let key: Key = Key {
            path: RepoPathBuf::from_string("foo".to_string()).unwrap(),
            hgid: crate::hgid::mocks::CS,
        };
        let orig = Orig(key.clone(), Some(key.clone()));
        let bytes = Bytes(key.clone(), Some(key.clone()));
        let hex = Hex(key.clone(), Some(key.clone()));

        let cbor_orig = encode(&orig).unwrap();
        let cbor_bytes = encode(&bytes).unwrap();
        let cbor_hex = encode(&hex).unwrap();

        assert_eq!(cbor_orig.len(), 113);
        assert_eq!(cbor_bytes.len(), 73);
        assert_eq!(cbor_hex.len(), 115);

        // Orig cannot decode bytes or hex.
        assert_eq!(&decode::<Orig>(&cbor_orig).unwrap().0, &key);
        decode::<Orig>(&cbor_bytes).unwrap_err();
        decode::<Orig>(&cbor_hex).unwrap_err();

        // Bytes can decode all 3 formats.
        assert_eq!(&decode::<Bytes>(&cbor_orig).unwrap().0, &key);
        assert_eq!(&decode::<Bytes>(&cbor_bytes).unwrap().0, &key);
        assert_eq!(&decode::<Bytes>(&cbor_hex).unwrap().0, &key);

        // Hex can decode all 3 formats.
        assert_eq!(&decode::<Hex>(&cbor_orig).unwrap().0, &key);
        assert_eq!(&decode::<Hex>(&cbor_bytes).unwrap().0, &key);
        assert_eq!(&decode::<Hex>(&cbor_hex).unwrap().0, &key);
    }
}
