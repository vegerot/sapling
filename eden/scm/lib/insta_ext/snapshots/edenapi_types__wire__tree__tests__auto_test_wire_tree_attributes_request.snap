---
source: tree.rs
expression: "WireTreeAttributesRequest::arbitrary(&mut g)"
---
{"5":true}
